kyuubi-server/src/main/scala/org/apache/kyuubi/parser/KyuubiParser.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.parser
+
+import scala.util.control.NonFatal
+
+import org.antlr.v4.runtime.{CharStream, CharStreams, CodePointCharStream, CommonTokenStream, IntStream}
+import org.antlr.v4.runtime.atn.PredictionMode
+import org.antlr.v4.runtime.misc.{Interval, ParseCancellationException}
+
+import org.apache.kyuubi.parser.plan.KyuubiNode
+import org.apache.kyuubi.sql.{KyuubiSqlBaseLexer, KyuubiSqlBaseParser, KyuubiSqlBaseParserBaseListener}
+
+class KyuubiParser {
+
+  val astBuilder = new KyuubiAstBuilder
+
+  def parsePlan(sqlText: String): KyuubiNode = parse(sqlText) { parser =>
+    astBuilder.visit(parser.singleStatement) match {
+      case plan: KyuubiNode => plan
+      case _ =>
+        throw new ParseException(s"Unsupported SQL statement $sqlText")
+    }
+  }
+
+  protected def parse[T](command: String)(toResult: KyuubiSqlBaseParser => T): T = {
+    val lexer = new KyuubiSqlBaseLexer(
+      new UpperCaseCharStream(CharStreams.fromString(command)))
+    lexer.removeErrorListeners()
+
+    val tokenStream = new CommonTokenStream(lexer)
+    val parser = new KyuubiSqlBaseParser(tokenStream)
+    parser.addParseListener(PostProcessor)
+    parser.removeErrorListeners()
+
+    try {
+      try {
+        // first, try parsing with potentially faster SLL mode
+        parser.getInterpreter.setPredictionMode(PredictionMode.SLL)
+        toResult(parser)
+      } catch {
+        case _: ParseCancellationException =>
+          // if we fail, parse with LL mode
+          tokenStream.seek(0) // rewind input stream
+          parser.reset()
+
+          // Try Again.
+          parser.getInterpreter.setPredictionMode(PredictionMode.LL)
+          toResult(parser)
+      }
+    } catch {
+      case NonFatal(t) =>
+        throw new ParseException(t.getMessage, Some(t))
+    }
+  }
+}
+
+case object PostProcessor extends KyuubiSqlBaseParserBaseListener {}
+
+/* Copied from Apache Spark's to avoid dependency on Spark Internals */
```

Comment:
what is this for? I only see the spark and those copiers hvie such an implementation, but rarely see it in other projects depending on antlr4

Reply:
agree with you, will remove this line

Reply:
I don't mean this line, I mean this class